newsrevenuehub/rev-engine
# DEV-2030 | refactor relationships involving organization

Please fill out each section even if it's just with "N/A"

#### Plan? (if this draft/incomplete indicate what you intend to do and how)

This PR is initially into DEV-2049 because it presupposes the migrations in that branch, and in short term, the limited diff from the PR will be clear. After DEV-2049 is merged, base for this PR should be set to main.

#### What's this PR do?

Rationalizes org model relationships (inclusive of child-grandchild relationships), as described in ticket and related Google doc.  At a high-level, an org should not be deletable if it has an RPs that point to DonationPages that have contributions. If no such contributions exist, deleting an org should cascade to referencing RevenuePrograms, DonationPages, and RoleAssignments.

Additionally, Plans referenced by one or more Organization cannot be deleted.

- Changes `Contribution.donation_page` on_delete behavior to PROTECT
- Changes `Organization.plan` on_delete behavior to PROTECT
- Changes `AbstractPage.revenue_program` on_delete behavior to CASCADE (this model is the base of Template and DonationPage)
- Changes `RoleAssignment.organization` on_delete behavior to CASCADE
- Adds unit tests for Organization, RevenueProgram, and Plan model around the business logic described above
- Updates some tests that broke when PROTECT behavior for contributions was introduced
- Updates the "follows" behavior in DonationPageAdmin.reversion_register to no longer follow contributions set. This was causing unexpected behavior when restoring an org or RP. Recovering the deleted RP led to manually deleted contributions being restored (because django-reversion doesn't take a new snapshot before deletion). 

#### Why are we doing this? How does it help us?

Makes Org-related model relationships fit business logic.

#### How should this be manually tested?

Setup:

- There should be an Org with a RP with a DonationPage with at least one contribution
- There should be a RoleAssignment referencing the Org
- The Org should have a plan, and there should be at least two plans

In the Django admin, try to delete the org. You should be blocked, and the related RP, DP and RA should still exist

Delete any contributions that point to any donation pages that point to any RPs owned by the org.  Now try again to delete the Org. This time it should work. Additionally, related RPs, DPs, and RAs should also be deleted.

Finally, confirm the Org/Plan relationship by trying to delete the Org's plan from the Django admin. You should be blocked. Go to the org and change its plan to something else (and you'll also need to do this for any other orgs that reference the plan). Now you should be able to delete the plan.


#### Have automated unit tests been added?

Yes

#### How should this change be communicated to end users?

N/A

#### Are there any smells or added technical debt to note?

No

#### Has this been documented? If so, where?

N/A

#### What are the relevant tickets? Add a link to any relevant ones.

[DEV-2030 | Rationalize Relationships Involving Organization model](https://news-revenue-hub.atlassian.net/jira/software/c/projects/DEV/boards/3?modal=detail&selectedIssue=DEV-2030)
[DEV-1581 | Rationalize Model Relationships](https://news-revenue-hub.atlassian.net/jira/software/c/projects/DEV/boards/3?modal=detail&selectedIssue=DEV-1581)

#### Do any changes need to be made before deployment to production (adding environment variables, for example)?

No

#### Are there next steps? If so, what? Have tickets been opened for them?

No
